python/pyarrow/_dataset.pyx

```diff
@@ -276,55 +276,68 @@ cdef class FileSystemDataset(Dataset):
 
     Parameters
     ----------
+    paths_or_selector : Union[FileSelector, List[FileInfo]]
+        List of files/directories to consume.
     schema : Schema
         The top-level schema of the DataDataset.
-    root_partition : Expression
-        The top-level partition of the DataDataset.
-    file_format : FileFormat
+    format : FileFormat
         File format to create fragments from, currently only
         ParquetFileFormat and IpcFileFormat are supported.
     filesystem : FileSystem
         The filesystem which files are from.
-    paths_or_selector : Union[FileSelector, List[FileInfo]]
-        List of files/directories to consume.
-    partitions : List[Expression]
+    partitions : List[Expression], optional
         Attach aditional partition information for the file paths.
+    root_partition : Expression, optional
+        The top-level partition of the DataDataset.
     """
 
     cdef:
         CFileSystemDataset* filesystem_dataset
 
-    def __init__(self, Schema schema not None, Expression root_partition,
-                 FileFormat file_format not None,
-                 FileSystem filesystem not None,
-                 paths_or_selector, partitions):
+    def __init__(self, paths_or_selector, schema=None, format=None,
```

Comment:
@jorisvandenbossche I think you can leave the type annotations here, just the `not None` part needs to be removed.

Reply:
The problem is then that it are positional keywords to cython. And although you can call it as keyword arg, you get all kinds of uninformative error messages once you do something wrong.

(maybe there is a better way to do this with cython, but didn't directly figure out. There is also not much docs about this)

Reply:
Like "TypeError: __init__() takes exactly 6 positional arguments (1 given)" when using all keyword arguments, but having a typo in one of them.

This was actually the reason that I first thought that you needed to pass everything as positional arguments, because I used `format=` instead of `file_format=`, and then you get such error. And so I wanted to solve that in this PR to allow keyword arguments, to find out that the cause was actually a wrong name and that keyword arguments were already supported .. :)

Reply:
Indeed, that's annoying. Do you wan't to report this upstream?

Reply:
Yes, I can do that. Need to make a small reproducer then, will put that on my to do list.

Reply:
@jorisvandenbossche schema, format and filesystem arguments are mandatory, please remove the default values.

Reply:
I know it might be confusing from the signature, but I did this to ensure proper error messages (and the docstring indicates which are optional, and you will get a proper error message when not passing a value)